File: frontend/src/app/models/birthday.model.ts
export interface User {
  id: number;
  email: string;
  fullName: string;
  bio?: string;
  avatarUrl?: string;
  status: string;
  roles?: string[];
  plan?: string;
  badges?: string[];
  lastAiWishGeneration?: string;
  lastAiGiftGeneration?: string;
  coins: number;
}

export type BirthdayCategory = 'Family' | 'Friend' | 'Work' | 'Other';

export interface Birthday {
  id: number;
  userId: number;
  name: string;
  birthdate: string; // ISO format 'YYYY-MM-DD'
  category: BirthdayCategory;
  photoUrl?: string;
  notes?: string;
  reminderDays: number;
  showAge?: boolean;
  email?: string;
  whatsapp?: string;
  gender?: 'Masculin' | 'Féminin' | 'Autre';
  isDeleted: boolean;
  interests?: string[];
  isFavorite?: boolean;
  shareToken?: string;
  // Party Details
  partyDate?: string;
  partyTime?: string;
  partyLocation?: string;
  partyDescription?: string;
  partyTasks?: PartyTask[];
  memories?: MemoryItem[];
  signatures?: ECardSignature[];
}

export interface PartyTask {
  id: number;
  birthdayId: number;
  description: string;
  assigneeName?: string;
  isCompleted: boolean;
  createdAt: string;
}

export interface MemoryItem {
  id: number;
  guestName: string;
  guestSessionId: string;
  message?: string;
  mediaUrl?: string;
  mediaType?: string;
  createdAt: string;
}

export interface ECardSignature {
  id: number;
  guestName: string;
  guestSessionId: string;
  message: string;
  color: string;
  fontFamily: string;
  createdAt: string;
}

export interface GiftSuggestion {
  name: string;
  estimatedPrice: string;
  whereToBuy: string;
  purchaseLink: string;
  imageUrl?: string;
  preparationTips: string;
}

export interface Gift {
  id: number;
  birthdayId: number;
  name: string;
  description: string;
  priceRange: string;
  url: string;
  imageUrl?: string;
  isReserved: boolean;
  reservedByName?: string;
  createdAt: string;
  // Community Features
  upvotes?: number;
  downvotes?: number;
  userVote?: 'UP' | 'DOWN' | null;
}

export interface SharedBirthday {
  id: number;
  name: string;
  birthdate: string;
  showAge: boolean;
  gender?: 'Masculin' | 'Féminin' | 'Autre';
  gifts: Gift[];
  // Party Details
  partyDate?: string;
  partyTime?: string;
  partyLocation?: string;
  partyDescription?: string;
  partyTasks?: PartyTask[];
  memories?: MemoryItem[];
  signatures?: ECardSignature[];
}

export function getZodiacSign(birthdate: string): { name: string; emoji: string } {
  if (!birthdate) return { name: '', emoji: '' };
  
  const date = new Date(birthdate);
  const day = date.getDate();
  const month = date.getMonth() + 1; // 1-12

  if ((month === 3 && day >= 21) || (month === 4 && day <= 19)) return { name: 'Bélier', emoji: '♈' };
  if ((month === 4 && day >= 20) || (month === 5 && day <= 20)) return { name: 'Taureau', emoji: '♉' };
  if ((month === 5 && day >= 21) || (month === 6 && day <= 20)) return { name: 'Gémeaux', emoji: '♊' };
  if ((month === 6 && day >= 21) || (month === 7 && day <= 22)) return { name: 'Cancer', emoji: '♋' };
  if ((month === 7 && day >= 23) || (month === 8 && day <= 22)) return { name: 'Lion', emoji: '♌' };
  if ((month === 8 && day >= 23) || (month === 9 && day <= 22)) return { name: 'Vierge', emoji: '♍' };
  if ((month === 9 && day >= 23) || (month === 10 && day <= 22)) return { name: 'Balance', emoji: '♎' };
  if ((month === 10 && day >= 23) || (month === 11 && day <= 21)) return { name: 'Scorpion', emoji: '♏' };
  if ((month === 11 && day >= 22) || (month === 12 && day <= 21)) return { name: 'Sagittaire', emoji: '♐' };
  if ((month === 12 && day >= 22) || (month === 1 && day <= 19)) return { name: 'Capricorne', emoji: '♑' };
  if ((month === 1 && day >= 20) || (month === 2 && day <= 18)) return { name: 'Verseau', emoji: '♒' };
  if ((month === 2 && day >= 19) || (month === 3 && day <= 20)) return { name: 'Poissons', emoji: '♓' };

  return { name: '', emoji: '' };
}

export interface BirthdayStats {
  total: number;
  todayCount: number;
  thisMonthCount: number;
  next30DaysCount: number;
  categoryDistribution: { [key in BirthdayCategory]?: number };
}

File: mobile_flutter/lib/core/models/birthday_model.dart
class BirthdayModel {
  final int id;
  final String name;
  final DateTime birthdate;
  final String category;
  final String? photoUrl;
  final String? notes;
  final int reminderDays;
  final bool showAge;
  final String? email;
  final String? whatsapp;
  final String? gender;
  final List<String> interests;
  final bool isFavorite;
  final String? shareToken;
  final String? partyDate;
  final String? partyTime;
  final String? partyLocation;
  final String? partyDescription;
  final DateTime? createdAt;

  BirthdayModel({
    required this.id,
    required this.name,
    required this.birthdate,
    this.category = 'AMIS',
    this.photoUrl,
    this.notes,
    this.reminderDays = 3,
    this.showAge = true,
    this.email,
    this.whatsapp,
    this.gender,
    this.interests = const [],
    this.isFavorite = false,
    this.shareToken,
    this.partyDate,
    this.partyTime,
    this.partyLocation,
    this.partyDescription,
    this.createdAt,
  });

  /// Days remaining until next birthday occurrence
  int get daysRemaining {
    final now = DateTime.now();
    var nextBirthday = DateTime(now.year, birthdate.month, birthdate.day);
    if (nextBirthday.isBefore(now) || nextBirthday.isAtSameMomentAs(now)) {
      nextBirthday = DateTime(now.year + 1, birthdate.month, birthdate.day);
    }
    return nextBirthday.difference(now).inDays;
  }

  /// Age on next birthday
  int get nextAge {
    final now = DateTime.now();
    var nextBirthday = DateTime(now.year, birthdate.month, birthdate.day);
    if (nextBirthday.isBefore(now)) {
      nextBirthday = DateTime(now.year + 1, birthdate.month, birthdate.day);
    }
    return nextBirthday.year - birthdate.year;
  }

  /// Whether the birthday is today
  bool get isToday {
    final now = DateTime.now();
    return now.month == birthdate.month && now.day == birthdate.day;
  }

  factory BirthdayModel.fromJson(Map<String, dynamic> json) {
    return BirthdayModel(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      birthdate: json['birthdate'] != null
          ? DateTime.parse(json['birthdate'])
          : DateTime.now(),
      category: json['category'] ?? 'AMIS',
      photoUrl: json['photoUrl'],
      notes: json['notes'],
      reminderDays: json['reminderDays'] ?? 3,
      showAge: json['showAge'] ?? true,
      email: json['email'],
      whatsapp: json['whatsapp'],
      gender: json['gender'],
      interests: json['interests'] != null
          ? List<String>.from(json['interests'])
          : [],
      isFavorite: json['isFavorite'] ?? false,
      shareToken: json['shareToken'],
      partyDate: json['partyDate'],
      partyTime: json['partyTime'],
      partyLocation: json['partyLocation'],
      partyDescription: json['partyDescription'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'birthdate': '${birthdate.year}-${birthdate.month.toString().padLeft(2, '0')}-${birthdate.day.toString().padLeft(2, '0')}',
      'category': category,
      'photoUrl': photoUrl,
      'notes': notes,
      'reminderDays': reminderDays,
      'showAge': showAge,
      'email': email,
      'whatsapp': whatsapp,
      'gender': gender,
      'interests': interests,
      'isFavorite': isFavorite,
      'partyDate': partyDate,
      'partyTime': partyTime,
      'partyLocation': partyLocation,
      'partyDescription': partyDescription,
    };
  }

  BirthdayModel copyWith({
    int? id,
    String? name,
    DateTime? birthdate,
    String? category,
    String? photoUrl,
    String? notes,
    int? reminderDays,
    bool? showAge,
    String? email,
    String? whatsapp,
    String? gender,
    List<String>? interests,
    bool? isFavorite,
    String? shareToken,
    String? partyDate,
    String? partyTime,
    String? partyLocation,
    String? partyDescription,
    DateTime? createdAt,
  }) {
    return BirthdayModel(
      id: id ?? this.id,
      name: name ?? this.name,
      birthdate: birthdate ?? this.birthdate,
      category: category ?? this.category,
      photoUrl: photoUrl ?? this.photoUrl,
      notes: notes ?? this.notes,
      reminderDays: reminderDays ?? this.reminderDays,
      showAge: showAge ?? this.showAge,
      email: email ?? this.email,
      whatsapp: whatsapp ?? this.whatsapp,
      gender: gender ?? this.gender,
      interests: interests ?? this.interests,
      isFavorite: isFavorite ?? this.isFavorite,
      shareToken: shareToken ?? this.shareToken,
      partyDate: partyDate ?? this.partyDate,
      partyTime: partyTime ?? this.partyTime,
      partyLocation: partyLocation ?? this.partyLocation,
      partyDescription: partyDescription ?? this.partyDescription,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}
